envoygateway dev environment install

eguzki opened the pull request:
### What

Epic issue https://github.com/Kuadrant/kuadrant-operator/issues/325

Deploy kuadrant on k8s cluster powered by EnvoyGateway

```
make local-setup GATEWAYAPI_PROVIDER=envoygateway
```

Let's hit that gateway (the steps only work on linux)

```sh
INGRESS_HOST=$(kubectl get gtw eg -n envoy-gateway-system -o jsonpath='{.status.addresses[0].value}')
INGRESS_PORT=$(kubectl get gtw eg -n envoy-gateway-system -o jsonpath='{.spec.listeners[?(@.name=="http")].port}')
```

```sh
curl --verbose --resolve www.example.com:${INGRESS_PORT}:${INGRESS_HOST} http://www.example.com:${INGRESS_PORT}/get
```
The expected response is 404 (no route yet)

```
* Added www.example.com:80:172.18.0.252 to DNS cache
* Hostname www.example.com was found in DNS cache
*   Trying 172.18.0.252:80...
* Connected to www.example.com (172.18.0.252) port 80 (#0)
> GET /get HTTP/1.1
> Host: www.example.com
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< date: Fri, 12 Apr 2024 09:28:19 GMT
< content-length: 0
< 
* Connection #0 to host www.example.com left intact
```

> Note that the target branch of this PR is not `main`, instead, it is `envoygateway` to collect all required changes to support entirely envoygateway gatewayapi provider

adam-cattermole commented on config/dependencies/envoy-gateway/gateway/gateway.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: eg
```
When we eventually merge the overarching feature we might want to see if we can deploy both `istio-ingressgateway` and envoy gateway `eg` with the same name/namespace so that all of the examples/quickstarts "just work" for both.

eguzki replied:
yeah. we might need to pick a generic name for both. It cannot be either `istio-ingressgateway` or `eg`.

adam-cattermole commented on make/envoy-gateway.mk:
```diff
@@ -0,0 +1,42 @@
+# egctl tool
+EGCTL=$(PROJECT_PATH)/bin/egctl
+EGCTL_VERSION ?= v1.0.0
+OS = linux
```
Should we get this from the system instead? `$(uname -s)` or something?

eguzki replied:
added automatic detection of OS and ARCH. EG has a custom bash script to do egctl installation: `https://gateway.envoyproxy.io/get-egctl.sh`. Unfortunately, currently it has some limitations, like it requires `sudo`. So, until they enhance their script, we can use our own.

adam-cattermole reviewed (approved):
Looks good just some minor comments

eguzki commented:
@adam-cattermole you approved the PR, however afterwards I added the automatic detection of OS and ARCH. Additionally, I rebased to resolve existing conflicts with current HEAD of `main`. Technically, a new approval is needed. Thanks.

adam-cattermole commented:
Sorry had a brief look at this and forgot to get back to it.

adam-cattermole commented on make/envoy-gateway.mk:
```diff
@@ -0,0 +1,58 @@
+EGCTL=$(PROJECT_PATH)/bin/egctl
+EGCTL_VERSION ?= v1.0.0
+OS = $(shell uname -s | tr '[:upper:]' '[:lower:]')
+UNAME_P := $(shell uname -p)
+ifeq ($(UNAME_P),x86_64)
+	ARCH = amd64
+endif
+ifeq ($(UNAME_P),aarch64)
+	ARCH = arm64
+endif
+ifneq ($(filter armv5%,$(UNAME_P)),)
+    ARCH = armv5
+endif
+ifneq ($(filter armv6%,$(UNAME_P)),)
+    ARCH = armv6
+endif
+ifneq ($(filter armv7%,$(UNAME_P)),)
+    ARCH = arm
+endif
```
Any reason not to use `uname -m` the same way it's done in the `get-egctl.sh`?

https://github.com/envoyproxy/gateway/blob/main/tools/hack/get-egctl.sh#L12-L25

On my system:
```
$ uname -p
unknown
```

eguzki replied:
from `uname` manpage,
```
       -p, --processor
              print the processor type (non-portable)
       -s, --kernel-name
              print the kernel name
       -m, --machine
              print the machine hardware name
```

they all give `x86_64` in my machine.

using `uname (GNU coreutils) 9.4`

> I also just noticed that the ARCH var is already set in alerts.mk so it's currently getting set regardless.

good catch. I will refactor a bit.

adam-cattermole replied:
Can't say for sure what the difference between `-p` and `-m` is in my env.
```
$ uname --version
uname (GNU coreutils) 9.5

$ uname -p
unknown

$ uname -m
x86_64
```

eguzki replied:
done, can you try getting egctl again?

adam-cattermole replied:
Nice yep all looks good now!

eguzki commented on make/alerts.mk:
```diff
@@ -1,12 +1,9 @@
 ##@ Alerts
-export WORKDIR ?= $(shell pwd)
 export IMAGE ?= quay.io/prometheus/prometheus
-export AVAILABILITY_SLO_RULES ?= $(WORKDIR)/examples/alerts/slo-availability.yaml
```
@R-Lawton you are the original owner of this script. heads up as I introduced small changes.

on a different note, why exporting makefile variables? are they being used in sub-processes? If not, they are not required.
